source/particle/property/crust_and_lithosphere_formation.cc

```diff
@@ -0,0 +1,216 @@
+            Assert(particle.get_surrounding_cell() == inputs.current_cell,
+                   ExcMessage("All particles must be in the same cell."));
+
+            material_inputs.position[p] = particle.get_location();
+            material_inputs.temperature[p] = inputs.solution[p][this->introspection().component_indices.temperature];
+            material_inputs.pressure[p] = inputs.solution[p][this->introspection().component_indices.pressure];
+
+            for (unsigned int d = 0; d < dim; ++d)
+              material_inputs.velocity[p][d] = inputs.solution[p][this->introspection().component_indices.velocities[d]];
+
+            for (unsigned int n = 0; n < this->n_compositional_fields(); ++n)
+              material_inputs.composition[p][n] = inputs.solution[p][this->introspection().component_indices.compositional_fields[n]];
+
+            material_inputs.composition[p][basalt_index] = particle.get_properties()[this->data_position];
+            material_inputs.composition[p][harzburgite_index] = particle.get_properties()[this->data_position+1];
+
+            Tensor<2,dim> grad_u;
+            for (unsigned int d=0; d<dim; ++d)
+              grad_u[d] = inputs.gradients[p][this->introspection().component_indices.velocities[d]];
+
+            material_inputs.strain_rate[p] = symmetrize (grad_u);
+
+            // We assume that the only reactions come from the
+            // crust and lithosphere formation reaction model.
+            material_outputs.reaction_terms[p][basalt_index] = 0.0;
+            material_outputs.reaction_terms[p][harzburgite_index] = 0.0;
+
+            ++p;
+          }
+
+        crust_lithosphere_formation->calculate_reaction_terms(material_inputs,
+                                                              material_outputs);
+
+        p = 0;
+        for (auto &particle: particles)
+          {
+            particle.get_properties()[this->data_position] += material_outputs.reaction_terms[p][basalt_index];
+            particle.get_properties()[this->data_position+1] += material_outputs.reaction_terms[p][harzburgite_index];
+            ++p;
+          }
+      }
+
+
+
+      template <int dim>
+      InitializationModeForLateParticles
+      CrustLithosphereFormation<dim>::late_initialization_mode () const
+      {
+        return interpolate_respect_boundary;
+      }
+
+
+
+      template <int dim>
+      UpdateTimeFlags
+      CrustLithosphereFormation<dim>::need_update() const
+      {
+        return update_time_step;
+      }
+
+
+
+      template <int dim>
+      UpdateFlags
+      CrustLithosphereFormation<dim>::get_needed_update_flags () const
+      {
+        return update_values | update_gradients;
+      }
+
+
+
+      template <int dim>
+      std::vector<std::pair<std::string, unsigned int>>
+      CrustLithosphereFormation<dim>::get_property_information() const
+      {
+        // TODO: This is the name we use in the input file. Does that make sense?
```

Comment:
hmm, it might conflict with the compositional field names which are also required to be basalt and harzburgite. Maybe "basalt_particle" and "harzburgite_particle"?